core/invoker/src/main/scala/org/apache/openwhisk/core/invoker/InvokerReactive.scala

```diff
@@ -145,43 +146,34 @@ class InvokerReactive(
     new MessageFeed("activation", logging, consumer, maxPeek, 1.second, processActivationMessage)
   })
 
-  /** Sends an active-ack. */
   private val ack: InvokerReactive.ActiveAck = (tid: TransactionId,
                                                 activationResult: WhiskActivation,
                                                 blockingInvoke: Boolean,
                                                 controllerInstance: ControllerInstanceId,
                                                 userId: UUID,
-                                                isSlotFree: Boolean) => {
+                                                acknowledegment: AcknowledegmentMessage) => {
```

Comment:
If the passed `acknowledegment` is a combined completion and result message or result message, it likely contains a `WhiskActivation` instance. So it duplicates the `activationResult` parameter which is only used to emit user metrics.

Can't we drop the `activationResult` parameter and change the condition for sending user metrics? If the `acknowledegment.result` is non-empty and a `Right(WhiskActivation)`, we emit metrics - otherwise not.

Reply:
yes - i considered doing that but because the current types allow for an `Either`, whereas all the call sites do know the activation, it would allow for a weakening should someone construct an ack with a Left where a Right is expected. I don't think there's a way to restrict the constructors (given they are case classes) without adding more types. So I think the current signature is tighter.